source/agora/node/EnrollmentManager.d

```diff
@@ -0,0 +1,138 @@
+import agora.common.crypto.Key;
+import agora.common.Hash;
+import agora.common.Types;
+import agora.common.crypto.Schnorr : Pair;
+
+import std.random;
+import std.stdio;
+
+/*******************************************************************************
+    Defines data for enrollment process
+*******************************************************************************/
+
+public struct EnrollmentData{
+    /// K: UTXO key, A public key matching a frozen UTXO
+    public PublicKey K;
+
+    /// X: random seed, The nth image of random value
+    public Hash X;
+
+    /// n: cycle length, A number within bounds [0; 1008]
+    /// original spec: [0; (Balance(K) % 40,000) * freezing period * 2]
+    public ulong n;
+
+    /// R: signature noise, The initial nonce used for signing
+    public Point R;
+
+    /// S: A signature for the messge H(K, X, n, R) and the key K, using R
+    public Signature S;
+
+    /***************************************************************************
+        Implements hashing support
+        Params:
+            dg = hashing function
+    ***************************************************************************/
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
+    {
+        hashPart(this.K, dg);
+        hashPart(this.X, dg);
+        hashPart(this.n, dg);
+        hashPart(this.R, dg);
+    }
+}
+
+/*******************************************************************************
+    Handles data and process for enrollment
+*******************************************************************************/
+
+public struct EnrollmentManager
+{
+    /// Node's key pair
+    private KeyPair node_keypair;
+
+    /// Random key for enrollment
+    private Pair random_key;
+
+    /// Node balance
+    private Amount balance;
+
+    // Random object
+    private Random genint;
+
+    /// EnrollmentData
+    private EnrollmentData* data;
+
+    /***************************************************************************
+        Constructor
+        Params:
+            keypair = the keypair of the owner node
+            balance = account balance of the owner node
+    ***************************************************************************/
+    public this(KeyPair keypair, Amount balance)
+    {
+        this.node_keypair = keypair;
+        this.balance = balance;
+        this.genint = Random(123);
+        this.data = null;
+    }
+
+    /***************************************************************************
+        Makes a data object for enrollment process
+        Returns:
+            a EnrollmentData object
+    ***************************************************************************/
+    public EnrollmentData getEnrollmentData()
+    {
+        this.data = new EnrollmentData();
+        
+        // public key
+        this.data.K = this.node_keypair.address;
+
+        // cycle length
+        this.data.n = uniform(0, 1008, this.genint);
```

Comment:
The number of voting rounds the node participates in should not be random. It should be set by the node operator.

Currently the Yellow paper says:

> 3. Enrollment process
> n (cycle length): the number of rounds a validator will participate in (currently fixed to (freezing period / 2)];

For now, I suggest making N a constructor parameter for `EnrollmentManager`.

Reply:
Thank you for your comment.
I'll change the number as described in the yellow paper, but I'd like to set the number inside the class not by the parameter of the constrcutor because the setting the number can be a burden to other classes such as the Node class. By a burden, I mean determining which number should be set. 
Any opinion about this would be helpful.

Reply:
Ok, you can keep this internal for now.